### PR TITLE
[MGDOBR-398] - Prevent the user from making concurrent modifications …

### PR DESCRIPTION
**Pull request by vaibhavjainwiz**
…to resources

[Insert JIRA Issue number and update link](https://issues.redhat.com/browse/MGDOBR-)
Jira issue : https://issues.redhat.com/browse/MGDOBR-398

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [x] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [x] All new functionality is tested
- [x] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.

</details>

### CODE REVIEW

File: manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java [hunk 1]
```diff
@@ -193,4 +197,35 @@ public void getBridge() {
...
+        assertThat(retrievedBridge.getStatus()).isEqualTo(ManagedResourceStatus.DELETED);
+    }
+
+    @Test
+    public void testDeleteBridge_whenStatusIsNotReady() {
+        Bridge bridge = createPersistBridge(ManagedResourceStatus.PROVISIONING);
+        assertThatExceptionOfType(BridgeLifecycleException.class).isThrownBy(() -> bridgesService.deleteBridge(bridge.getId(), bridge.getCustomerId()));
+    }
+
+    private Bridge createPersistBridge(ManagedResourceStatus status) {
```
Comment (robpblake) [inline manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java:221]: @vaibhavjainwiz Take a look at `Fixtures.createBridge()`. It should give you most of what you need here.

File: manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java [hunk 2]
```diff
@@ -193,4 +197,35 @@ public void getBridge() {
...
+        bridgesService.createBridge(TestConstants.DEFAULT_CUSTOMER_ID, request);
+
+        //Wait for Workers to complete
+        Bridge bridge = TestUtils.waitForBridgeToBeReady(bridgesService);
+
+        bridgesService.deleteBridge(bridge.getId(), bridge.getCustomerId());
+
+        Bridge retrievedBridge = bridgesService.getBridge(bridge.getId(), TestConstants.DEFAULT_CUSTOMER_ID);
+        assertThat(retrievedBridge.getStatus()).isEqualTo(ManagedResourceStatus.DELETED);
+    }
```
Comment (r00ta) [inline manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java:213]: I think I'm missing a bit why this test works. `bridgesService.deleteBridge` should mark the Bridge with `DEPROVISION`

Comment (manstis) [reply manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java:213]: I agree with @r00ta unless the Shard is running the `Bridge` should only be in `DEPROVISION` status. L#206 will wait for the `WorkManager` to execute the `Work` to provision the `Bridge` but its `status` will remain `ACCEPTED` and its `dependencyStatus` will be `READY` (as this test is annotated with `@TestProfile(WorkerSchedulerProfile.class)`. However nothing should move the `Bridge` to `DELETED` in the `manager` module.

File: manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java (hunk 1 above)
Comment (vaibhavjainwiz) [reply manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java:218]: done

File: manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java (hunk 2 above)
Comment (vaibhavjainwiz) [reply manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java]: testcase is updated. Could you please review it again.

File: manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java [hunk 3]
```diff
@@ -187,6 +188,10 @@ public void deleteProcessor(String bridgeId, String processorId, String customer
         if (processor == null) {
             throw new ItemNotFoundException(String.format("Processor with id '%s' does not exist on bridge '%s' for customer '%s'", processorId, bridgeId, customerId));
         }
+        if (!isProcessorDeletable(processor)) {
+            throw new ProcessorLifecycleException("Processor could only be deleted if its in READY/FAILED state.");
+        }
+        processor.setStatus(ManagedResourceStatus.DEPROVISION);
```
Comment (r00ta) [inline manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java:194]: You can remove this statement I think

Comment (manstis) [inline manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java:194]: This is done here [too](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/c6eff16620815d47f74ada6483ad18d14256ca20/manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java#L197). Which do you want to keep?

File: manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java [hunk 4]
```diff
@@ -124,6 +124,11 @@ public void deleteBridge(String id, String customerId) {
         }
 
         Bridge bridge = findByIdAndCustomerId(id, customerId);
+        if (!isBridgeDeletable(bridge)) {
+            throw new BridgeLifecycleException("Bridge could only be deleted if its in READY/FAILED state.");
+        }
+        bridge.setStatus(ManagedResourceStatus.DEPROVISION);
```
Comment (r00ta) [inline manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java:130]: You can remove this statement

Comment (manstis) [inline manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java:130]: This is done here [too](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/c6eff16620815d47f74ada6483ad18d14256ca20/manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java#L134). Which do you want to keep?

File: manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java [hunk 5]
```diff
@@ -193,4 +198,27 @@ public void getBridge() {
     public void getBridge_bridgeDoesNotExist() {
         assertThatExceptionOfType(ItemNotFoundException.class).isThrownBy(() -> bridgesService.getBridge("foo"));
     }
+
+    @Test
+    public void testDeleteBridge() {
```
Comment (manstis) [inline manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java:203]: You could add a test for the the `Bridge` is `FAILED` too.

File: manager/src/test/java/com/redhat/service/bridge/manager/ProcessorServiceTest.java [hunk 6]
```diff
@@ -366,12 +376,9 @@ public void testGetProcessorsCount() {
     @Test
     public void testDeleteProcessor() {
```
Comment (manstis) [inline manager/src/test/java/com/redhat/service/bridge/manager/ProcessorServiceTest.java:377]: You could add a test for the the `Processor` is `FAILED` too.

File: manager/src/test/java/com/redhat/service/bridge/manager/api/internal/ShardBridgesSyncAPITest.java [hunk 7]
```diff
@@ -248,6 +248,9 @@ public void testGetBridgesToDelete() {
 
         BridgeDTO bridge = bridgesToDeployOrDelete.get(0);
 
+        bridge.setStatus(ManagedResourceStatus.READY);
```
Comment (manstis) [inline manager/src/test/java/com/redhat/service/bridge/manager/api/internal/ShardBridgesSyncAPITest.java:251]: The `Bridge` should be in `READY` state after [this](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/c6eff16620815d47f74ada6483ad18d14256ca20/manager/src/test/java/com/redhat/service/bridge/manager/api/internal/ShardBridgesSyncAPITest.java#L242) completes.

File: manager/src/test/java/com/redhat/service/bridge/manager/api/internal/ShardBridgesSyncAPITest.java [hunk 8]
```diff
@@ -301,6 +304,9 @@ public void testNotifyDeletion() {
...
```
Comment (manstis) [inline manager/src/test/java/com/redhat/service/bridge/manager/api/internal/ShardBridgesSyncAPITest.java:307]: The `Bridge` should be in `READY` state when [this](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/c6eff16620815d47f74ada6483ad18d14256ca20/manager/src/test/java/com/redhat/service/bridge/manager/api/internal/ShardBridgesSyncAPITest.java#L298) completes.

File: openapi/openapi.yaml [hunk 9]
```diff
@@ -11,39 +11,39 @@ info:
     url: https://www.apache.org/licenses/LICENSE-2.0.html
   version: 0.0.1
 tags:
-- name: Bridges
-  description: "The API that allow the user to retrieve, create or delete Bridge instances."
-- name: Error Catalog
-  description: List and get the error definitions from the error catalog.
-- name: Processors
-  description: "The API that allow the user to retrieve, create or delete Processors\
+  - name: Bridges
```
Comment (r00ta) [inline openapi/openapi.yaml:14]: I think you will have to revert the changes in this file, otherwise the pipeline will fail

File: manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java (hunk 3 above)
Comment (vaibhavjainwiz) [reply manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java:194]: done

File: manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java (hunk 4 above)
Comment (vaibhavjainwiz) [reply manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java:130]: done

Comment (robpblake) [reply manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java:130]: @vaibhavjainwiz We don't need this:

```
bridgeDAO.getEntityManager().merge(bridge)
```

The Bridge is already loaded into the `PersistenceContext`, so any changes will be automatically flushed to the DB when the transaction commits. Calling `bridge.setStatus()` is the correct thing to do.

File: manager/src/test/java/com/redhat/service/bridge/manager/api/internal/ShardBridgesSyncAPITest.java (hunk 7 above)
Comment (vaibhavjainwiz) [reply manager/src/test/java/com/redhat/service/bridge/manager/api/internal/ShardBridgesSyncAPITest.java:251]: Here we just create a bridge which set it to `APPROVED` state and then try to delete it. that why I set to `READY` manually so that we could test delete funcntionality.

File: manager/src/test/java/com/redhat/service/bridge/manager/api/internal/ShardBridgesSyncAPITest.java (hunk 8 above)
Comment (vaibhavjainwiz) [reply manager/src/test/java/com/redhat/service/bridge/manager/api/internal/ShardBridgesSyncAPITest.java:307]: Here we just create a bridge which set it to `APPROVED` state and then try to delete it. that why I set to `READY` manually so that we could test delete funcntionality.

File: manager/src/test/java/com/redhat/service/bridge/manager/ProcessorServiceTest.java (hunk 6 above)
Comment (manstis) [reply manager/src/test/java/com/redhat/service/bridge/manager/ProcessorServiceTest.java:377]: That is when it is `PROVISIONING` not `FAILED`. When `FAILED` you should be allowed to delete it, but I don't see a test for that... however please point me to one. I could not see it.

File: manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java (hunk 5 above)
Comment (manstis) [reply manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java:203]: That is when it is `PROVISIONING` not `FAILED`. When `FAILED` you should be allowed to delete it, but I don't see a test for that... however please point me to one. I could not see it.

File: manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java (hunk 4 above)
Comment (manstis) [reply manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java]: Yes, this is probably residue from when I merged the `delete(..)` and `doDelete(..)` methods :+1:

Comment (vaibhavjainwiz) [reply manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java]: @robpblake I am completely agree with you. I was also suprised to see this new changes but keep them to myself to think these changes are made for a reason. 
I will revert them

File: manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java (hunk 5 above)
Comment (vaibhavjainwiz) [reply manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java:203]: okies, I miss took your earlier comment. I thought you are asking to add `failure` usecase. I will add it.

File: manager/src/test/java/com/redhat/service/bridge/manager/ProcessorServiceTest.java (hunk 6 above)
Comment (vaibhavjainwiz) [reply manager/src/test/java/com/redhat/service/bridge/manager/ProcessorServiceTest.java:377]: okies, I miss took your earlier comment. I thought you are asking to add failure usecase. I will add it.

File: manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java (hunk 5 above)
Comment (vaibhavjainwiz) [reply manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java:203]: I had added new testcase for `FAILED` scenerio.
https://github.com/vaibhavjainwiz/sandbox/blob/f68de90bd03dc5bf56f0059025d09312f694e6b6/manager/src/test/java/com/redhat/service/bridge/manager/BridgesServiceTest.java#L213

File: manager/src/test/java/com/redhat/service/bridge/manager/ProcessorServiceTest.java (hunk 6 above)
Comment (vaibhavjainwiz) [reply manager/src/test/java/com/redhat/service/bridge/manager/ProcessorServiceTest.java:377]: I had new testcase for `FAILED` status
https://github.com/vaibhavjainwiz/sandbox/blob/f68de90bd03dc5bf56f0059025d09312f694e6b6/manager/src/test/java/com/redhat/service/bridge/manager/ProcessorServiceTest.java#L393

File: manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java (hunk 4 above)
Comment (vaibhavjainwiz) [reply manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java]: done

Comment (vaibhavjainwiz) [reply manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java]: I had removed other duplicate statement

File: manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java (hunk 3 above)
Comment (vaibhavjainwiz) [reply manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java]: I had removed other duplicate statement

**Review by manstis — APPROVED** [review]

**Review by robpblake — APPROVED** [review]
Thank you @vaibhavjainwiz . LGTM!

**Review by r00ta — APPROVED** [review]

**Review by sutaakar — APPROVED** [review]
Looks good to me
